Clicking on save button dispatches a MouseEvent, button location in element's saveButtonClientRect

squelart opened the pull request:
#### b9dad072710ae23d168919abf01468c1763b8d33
<pre>
Clicking on save button dispatches a MouseEvent, button location in element&apos;s saveButtonClientRect
<a href="https://bugs.webkit.org/show_bug.cgi?id=252967">https://bugs.webkit.org/show_bug.cgi?id=252967</a>
rdar://problem/105952126

Reviewed by Tim Nguyen.

Instead of sending a CustomEvent with some hand-picked details, a click on the save button now dispatches a full copy of the original MouseEvent, and the location of the (shadow) save button can be found the element&apos;s saveButtonClientRect property.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::saveButtonClientRect const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLAttachmentElement.idl:

Canonical link: <a href="https://commits.webkit.org/261101@main">https://commits.webkit.org/261101@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9644dcc2fc79d035e97dd59bfbe7ccfa73a824b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1768 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119308 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10629 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102627 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15574 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98761 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43796 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97535 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30426 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85655 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12137 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31764 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12720 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8736 "Found 1 new test failure: swipe/pushState-cached-back-swipe.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18085 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51379 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7692 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14575 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Looks good, would be nice to have test coverage

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -201,15 +202,18 @@ class AttachmentSaveEventListener final : public EventListener {
     {
         if (event.type() == eventNames().clickEvent) {
+            ASSERT(event.isMouseEvent());
+            auto& mouse = static_cast<MouseEvent&>(event);
```
```suggestion
            auto& mouseEvent = downcast<MouseEvent>(event);
```
downcast also covers the assert.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -201,15 +202,18 @@ class AttachmentSaveEventListener final : public EventListener {
         if (event.type() == eventNames().clickEvent) {
+            ASSERT(event.isMouseEvent());
+            auto& mouse = static_cast<MouseEvent&>(event);
+            auto copiedEvent = MouseEvent::create(m_attachment->attributeWithoutSynchronization(saveAttr()), mouse.bubbles() ? Event::CanBubble::Yes : Event::CanBubble::No,
```
No need to pass over CanBubble::Yes/No, you should decide whether you want your event to bubble.

nt1m replied:
(by default bubbling is off, so that might be a good default)

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -201,15 +202,18 @@ class AttachmentSaveEventListener final : public EventListener {
+            ASSERT(event.isMouseEvent());
+            auto& mouse = static_cast<MouseEvent&>(event);
+            auto copiedEvent = MouseEvent::create(m_attachment->attributeWithoutSynchronization(saveAttr()), mouse.bubbles() ? Event::CanBubble::Yes : Event::CanBubble::No,
+                mouse.cancelable() ? Event::IsCancelable::Yes : Event::IsCancelable::No, mouse.composed() ? Event::IsComposed::Yes : Event::IsComposed::No,
```
Afaik the event is not cancelable, `event.preventDefault()` from the save event wouldn't do anything here.

nt1m commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -201,15 +202,17 @@ class AttachmentSaveEventListener final : public EventListener {
+            auto copiedEvent = MouseEvent::create(
+                m_attachment->attributeWithoutSynchronization(saveAttr()), Event::CanBubble::No, Event::IsCancelable::No, Event::IsComposed::No,
+                mouseEvent.view(), mouseEvent.detail(), mouseEvent.screenX(), mouseEvent.screenY(), mouseEvent.clientX(), mouseEvent.clientY(),
+                mouseEvent.modifierKeys(), mouseEvent.button(), mouseEvent.buttons(), mouseEvent.syntheticClickType(), mouseEvent.relatedTarget());
```
relatedTarget() should probably be nullptr, see https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/relatedTarget

squelart committed:
Clicking on save button dispatches a MouseEvent, button location in element's saveButtonClientRect
https://bugs.webkit.org/show_bug.cgi?id=252967
rdar://problem/105952126

Reviewed by Tim Nguyen.

Instead of sending a CustomEvent with some hand-picked details, a click on the save button now dispatches a full copy of the original MouseEvent, and the location of the (shadow) save button can be found the element's saveButtonClientRect property.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::saveButtonClientRect const):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/HTMLAttachmentElement.idl:

Canonical link: https://commits.webkit.org/261101@main
